File: Solution/640.求解方程/640.求解方程.py
'''
Descroption: LeetCode 640. 求解方程
Author: EmoryHuang
Date: 2022-08-10 09:02:40
Method:
模拟
遍历字符串，将其转换为 k * x = val
'''


class Solution:
    def solveEquation(self, equation: str) -> str:
        factor = val = 0
        i, n, sign = 0, len(equation), 1
        while i < n:
            # 等号变号
            if equation[i] == '=':
                sign = -1
                i += 1
                continue

            # 加减号处理
            s = sign
            if equation[i] == '+':
                i += 1
            elif equation[i] == '-':
                s = -s
                i += 1

            # 数字处理
            num, flag = 0, False
            while i < n and equation[i].isdigit():
                flag = True
                num = num * 10 + int(equation[i])
                i += 1

            # 变量处理
            if i < n and equation[i] == 'x':
                factor += s * num if flag else s
                i += 1
            else:
                val += s * num
        if factor == 0:
            return "No solution" if val else "Infinite solutions"
        return "No solution" if val % factor else f"x={-val // factor}"

File: Solution/640.求解方程/640.求解方程.java
/*
 * @Descroption: LeetCode 640. 求解方程
 * @Author: EmoryHuang
 * @Date: 2022-08-10 09:09:31
 * @Method:
 * 模拟
 * 遍历字符串，将其转换为 k * x = val
 */

class Solution {

  public String solveEquation(String equation) {
    int factor = 0, val = 0;
    int i = 0, n = equation.length(), sign = 1;
    while (i < n) {
      // 等号变号
      if (equation.charAt(i) == '=') {
        sign = -1;
        i++;
        continue;
      }
      // 加减号处理
      int s = sign;
      if (equation.charAt(i) == '+') {
        i++;
      } else if (equation.charAt(i) == '-') {
        s = -s;
        i++;
      }
      // 数字处理
      int num = 0;
      boolean flag = false;
      while (i < n && equation.charAt(i) >= '0' && equation.charAt(i) <= '9') {
        flag = true;
        num = num * 10 + equation.charAt(i) - '0';
        i++;
      }
      // 变量处理
      if (i < n && equation.charAt(i) == 'x') {
        factor += flag ? s * num : s;
        i++;
      } else {
        val += s * num;
      }
    }
    if (factor == 0) return val == 0 ? "Infinite solutions" : "No solution";
    if (val % factor != 0) return "No solution";
    return "x=" + (-val / factor);
  }
}
